src/txrebroadcast.cpp

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <chainparams.h>
+#include <consensus/consensus.h>
+#include <logging.h>
+#include <miner.h>
+#include <script/script.h>
+#include <txrebroadcast.h>
+#include <util/time.h>
+#include <validation.h>
+
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+/** We rebroadcast 3/4 of max block weight to reduce noise due to circumstances
+ *  such as miners mining priority transactions. */
+static constexpr unsigned int MAX_REBROADCAST_WEIGHT{3 * MAX_BLOCK_WEIGHT / 4};
+
+/** Default minimum age for a transaction to be rebroadcast */
+static constexpr std::chrono::minutes REBROADCAST_MIN_TX_AGE{30min};
+
+/** Maximum number of times we will rebroadcast a transaction */
+static constexpr int MAX_REBROADCAST_COUNT{6};
+
+/** Minimum amount of time between returning the same transaction for
+ * rebroadcast */
+static constexpr std::chrono::hours MIN_REATTEMPT_INTERVAL{4h};
+
+/** The maximum number of entries permitted in m_attempt_tracker */
+static constexpr int MAX_ENTRIES{500};
```

Comment:
what is the fundamental resource that is being spent in this attack? are you evaluating the possibility of asymmetric bandwidth usage? if that's the case then you also need to factor in rolling over `filterInventoryKnown`. 

also I don't understand how the attacker avoids spending any funds- say A and A' are two versions of the same txn via RBF. A is sent to the victim & A' is sent to the rest of the network. according to the current patch, if A' gets mined into a block, A is removed from the attempt tracker.

Reply:
Thanks for the second clarifying update, effectively this is documented L124 in `src/txrebroadcast.cpp`, "No existing entry, we will rebroadcast, so create new one". 

So IIUC, even if `m_attempt_tracker` is fulfilled with pinning transactions, it won't block node rebroadcast capability. I think pinning costs of `m_attempt_tracker` roughly holds above but it can't be abused by an attacker to jam rebroadcasting, so we're good.

I would recommend to document this "pinning-doesn't-block-rebroadcast-so-not-a-new-tx-relay jamming-vector" around `m_attempt_tracker` declaration, quite important design aspect imo.

W.r.t to asymmetric bandwidth usage, I don't think `m_attempt_tracker` pinning is making it better or worst, I share the opinion we're ultimately protected by `filterInventoryKnown`.

>  I don't understand how the attacker avoids spending any funds- say A and A' are two versions of the same txn via RBF. A is sent to the victim & A' is sent to the rest of the network. according to the current patch, if A' gets mined into a block, A is removed from the attempt tracker.

I should have precise this point. Transaction A feerate to successfully pin `m_attempt_tracker` must be at 170 sat/vbyte. However transaction A' feerate just need to be above miners'mempools min feerates to block propagation of transaction A. If transaction A' feerate is 2 sat/vbyte, with same equation laid out above, effective pinning cost for an attacker is 12_000 * 82 * 2 = 1_968_000 sats and not 167_280_000 sats. Miners observe and eventually mine transactions with a far lower-feerate than the ones maliciously sent to non-miner nodes. This is what I meant by "not burnt if the mempool isolation is operated properly".

Reply:
> So IIUC, even if m_attempt_tracker is fulfilled with pinning transactions, it won't block node rebroadcast capability. 

yeah, exactly. 

> I would recommend to document this "pinning-doesn't-block-rebroadcast-so-not-a-new-tx-relay jamming-vector" around m_attempt_tracker declaration in src/txreconciliation.h, it's quite an important design aspect imo.

I don't think it makes sense to document what data structures _don't_ do. 

> I should have precise this point...

Ah I understand now, thanks for clarifying.